File: backend/src/main/java/com/gestionstock/backend/service/fournisseur/ClientService.java
package com.gestionstock.backend.service.fournisseur;

import com.gestionstock.backend.dto.fournisseur.ClientDTO;

import java.util.List;

/**
 * Interface du service Client.
 * Définit les opérations métier pour la gestion des clients.
 */
public interface ClientService {

    /** Récupère tous les clients */
    List<ClientDTO> getAllClients();

    /** Récupère un client par son identifiant */
    ClientDTO getClientById(Long id);

    /** Recherche les clients par nom ou prénom (recherche partielle) */
    List<ClientDTO> searchClients(String search);

    /** Crée un nouveau client */
    ClientDTO createClient(ClientDTO clientDTO);

    /** Met à jour un client existant */
    ClientDTO updateClient(Long id, ClientDTO clientDTO);

    /** Supprime un client par son identifiant */
    void deleteClient(Long id);
}

File: frontend/src/app/core/services/client.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { Client } from '../models/client.model';
import { API_BASE_URL } from './api.config';

const API_URL = API_BASE_URL;

/**
 * Service Angular pour la gestion des clients.
 * Communique avec le backend via les endpoints REST /api/clients.
 */
@Injectable({
  providedIn: 'root'
})
export class ClientService {

  private http = inject(HttpClient);

  /** Récupère tous les clients */
  getAllClients(): Observable<Client[]> {
    return this.http.get<Client[]>(`${API_URL}/clients`);
  }

  /** Récupère un client par son ID */
  getClientById(id: number): Observable<Client> {
    return this.http.get<Client>(`${API_URL}/clients/${id}`);
  }

  /** Recherche les clients par nom ou prénom */
  searchClients(query: string): Observable<Client[]> {
    return this.http.get<Client[]>(`${API_URL}/clients/search?q=${query}`);
  }

  /** Crée un nouveau client */
  createClient(client: Client): Observable<Client> {
    return this.http.post<Client>(`${API_URL}/clients`, client);
  }

  /** Met à jour un client existant */
  updateClient(id: number, client: Client): Observable<Client> {
    return this.http.put<Client>(`${API_URL}/clients/${id}`, client);
  }

  /** Supprime un client */
  deleteClient(id: number): Observable<void> {
    return this.http.delete<void>(`${API_URL}/clients/${id}`);
  }
}
